Add support for AWS Managed NodeGroups

metral opened the pull request:
<!--Thanks for your contribution. See [CONTRIBUTING](CONTRIBUTING.md)
    for Pulumi's contribution guidelines.

    Help us merge your changes more quickly by adding more details such
    as labels, milestones, and reviewers.-->

### Proposed changes

- feat(nodes): add eks.createManagedNodeGroup()

PR includes new examples/test: `examples/managed-nodegroups`.

---

**Support Details:**
(compared to original self-managed node groups)

* Security Groups
    * > Amazon EKS clusters beginning with Kubernetes version 1.14 and platform version eks.3 create a cluster security group as part of cluster creation (or when a cluster is upgraded to this Kubernetes version and platform version). This security group is designed to allow all traffic from the control plane and managed node groups to flow freely between each other
    * > Amazon EKS managed node groups are automatically configured to use the cluster security group.
    * Users cannot provide their own security group, or ingress rules for use with the cluster
    * Users can only provide their own security group for remote configuration for Node SSH access.
* SSH Key
    * Only key pair name can be provided.
* Cluster Attributes
    * Only k8s labels and AWS tags on the managed node groups are supported.
    * `kubelet-extra-args` is not supported.
        * Taints are not supported.
    * > AWS handles node updates and terminations to gracefully drain nodes
* Instance Type
    * TF allows `instance_types` (plural) to be a set, but EKS appears to only accept a single, string value.
        * The Pulumi support is mapped to a `pulumi.Output<string>`, which is inline with EKS support
        * The variable in Pulumi is `instanceTypes` (plural per the TF provider), even though its value is singular.
* AMI ID
    * > Instances in a managed node group use the latest version of the Amazon EKS-optimized Amazon Linux 2 AMI for its cluster's Kubernetes version. You can choose between standard and GPU variants of the Amazon EKS-optimized Amazon Linux 2 AMI.

References:

- https://docs.aws.amazon.com/eks/latest/userguide/create-managed-node-group.html
- https://docs.aws.amazon.com/eks/latest/userguide/managed-node-groups.html
- https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html
- https://www.terraform.io/docs/providers/aws/r/eks_node_group.html

### Related issues (optional)

Closes https://github.com/pulumi/pulumi-eks/issues/278

lukehoban commented on nodejs/eks/nodegroup.ts:
```diff
@@ -626,3 +626,237 @@ function tagsToAsgTags(tagsInput: InputTags): pulumi.Output<string> {
+
+    if (!args.instanceRole) {
+        throw new Error(`an instanceRole is required for the managed node group`);
+    }
```
Why is this needed instead of making this property required?

metral replied:
Good catch. Thanks

lukehoban commented on nodejs/eks/nodegroup.ts:
```diff
@@ -626,3 +626,237 @@ function tagsToAsgTags(tagsInput: InputTags): pulumi.Output<string> {
+    let maxSize: pulumi.Input<number> = 0;
+    if (!args.scalingConfig || args.scalingConfig.desiredCapacity === undefined) {
+        desiredCapacity = 2;
+    } else { desiredCapacity = args.scalingConfig.desiredCapacity; }
```
just for consistency we can format these as normal with new lines. 

And total nit - this whole thing could be `const desiredCapacity = args.scalingConfig && args.scalingConfig.desiredCapacity || 2`.

lukehoban commented on nodejs/eks/nodegroup.ts:
```diff
@@ -626,3 +626,237 @@ function tagsToAsgTags(tagsInput: InputTags): pulumi.Output<string> {
+    } else { maxSize = args.scalingConfig.maxSize; }
+
+    if (args.amiType === undefined) {
+        args.amiType = "AL2_x86_64";
```
Are these defaults (both here and above) not the same defaults that you would get if you pass undefined through?  If they are the same - why are we recomputing them here?  If they are not - how did we come up with these defaults?

metral replied:
Agreed we should not recompute. I'll pass undefined to lower maintenance

lukehoban commented on nodejs/eks/nodegroup.ts:
```diff
@@ -626,3 +626,237 @@ function tagsToAsgTags(tagsInput: InputTags): pulumi.Output<string> {
+ * ManagedNodeGroupBaseOptions represents the common configuration settings for
+ * AWS managed NodeGroups.
+ */
+export interface ManagedNodeGroupBaseOptions {
```
Is there a reason we need to define a new interface here vs using either our existing NodeGroup interface or the raw AWS input interface?  Is it that we are making some required properties optional?  Can we somehow ensure that if new properties get added to the underlying interface we automatically support them?

metral replied:
Great points. I'll try to refactor this to reuse `aws.eks.NodeGroup` instead to support the provider longterm.

metral commented:
> The one larger question - which I think I know the answer to but would love to hear your thoughts - do we need the createManagedNodeGroup instead of just allowing folks to use aws.eks.NodeGroup directly instead? What benefits do uses get from having the API available here?

Great point, I'll refactor.

> And are there enough benefits to non-managed NodeGroups that you expect we’ll still want them  here long term?

I believe we should keep self-managed NodeGroups in maintenance mode for now, and reassess later. Keeping it around allows:
- Existing clusters to have a path to use the AWS managed node groups if necessary, and
- If the cluster needs maximum configuration, this is the option best suited.

As the Support Details in the OP note, some users may want specificity not yet supported in AWS managed node groups, e.g.: `kubelet-extra-args`, Taints, custom security groups, and AMI version pinning.

metral commented:
> Instance Type
> * TF allows instance_types (plural) to be a set, but EKS appears to only accept a single, string value.
> * The Pulumi support is mapped to a pulumi.Output<string>, which is inline with EKS support
> * The variable in Pulumi is instanceTypes (plural per the TF provider), even though its value is singular.

^ Thoughts on support details from OP @stack72 ?

metral commented:
@joeduffy @lukehoban @stack72 PTAL

stack72 commented:
> > Instance Type
> > 
> > * TF allows instance_types (plural) to be a set, but EKS appears to only accept a single, string value.
> > * The Pulumi support is mapped to a pulumi.Output, which is inline with EKS support
> > * The variable in Pulumi is instanceTypes (plural per the TF provider), even though its value is singular.
> 
> ^ Thoughts on support details from OP @stack72 ?

@metral so the AWS CLI actually uses InstanceTypes as well - I think this could be for an upcoming feature - I think we should keep that

metral commented:
CI is ✅

PTAL @lukehoban @joeduffy

lukehoban commented on nodejs/eks/cluster.ts:
```diff
@@ -483,14 +485,55 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
+    *
+    * TODO(metral)
+    * When https://git.io/JeXom is resolved, we can move away from kubectl
+    * shell out.
```
Nit - in most other repos we've used the style `TODO[pulumi/pulumi-kubernetes#264]: Note about the issue.`

metral replied:
Done

lukehoban commented on nodejs/eks/cluster.ts:
```diff
@@ -483,14 +485,55 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
+    *   resource error message.
+    *
+    * For self-managed nodegroups:
+    *   We continue to always create the aws-auth configmap.
```
I don't see this implemented anywhere - we seem to have just deleted the `new k8s.core.v1.ConfigMap`.

Won't that be a breaking change for existing deployments? (do we have any update testing that would verify this doesn't break things?)

lblackstone replied:
Yeah, it looks like this change moves this CM out of Pulumi's management for both cases. I expect that we'd want to keep the existing way for non-managed node groups.

metral replied:
The pulumi managed ConfigMap was replaced with the `kubectl apply` - they both produce the same configmap result in the cluster, but it is now no longer under Pulumi management.

We must do this to satisfy both configmap scenarios when using self-managed nodegroups vs AWS managed.

This is not a breaking change as the CI runs with `RunUpdateTest` on all existing tests to update to this PR and all tests are green.

>  I expect that we'd want to keep the existing way for non-managed node groups

We cannot deduce this from the cluster `createCore()` (the orig location of the configmap creation), as the user has a couple of ways of creating self-managed node groups and using the new `createManagedNodeGroup()` API. Naturally it felt easier to replace this here with `kubectl apply`

lukehoban replied:
If I understand what will happen during an upgrade to a new version of this package when using existing node groups:
1. During the update, we will now do a `kubectl apply` which will in principle update the existing ConfigMap and make no changes.
2. At the end of the update, we will destroy the `${name}-nodeAccess` ConfigMap as it no longer exists in the Pulumi program, this will blow away the changes that were just made above.

Are you sure that the above does not happen?  If it does not - what does in practice happen during this update? (and how do we avoid the removal of the `new k8s.core.v1.ConfigMap` from destroying the ConfigMap?

metral replied:
I tried this on existing self-managed nodegroups, and you are correct that deleting the nodeAccess pulumi-managed CM removes the `aws-auth` file configmap after an update to the latest version of this PR.

lukehoban commented on nodejs/eks/cluster.ts:
```diff
@@ -483,14 +485,55 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
+        fs.writeFileSync(tmpKubeconfig.fd, JSON.stringify(kc));
+
+        // Compute the required aws-auth JSON manifest and dump it to a file.
+        fs.writeFileSync(tmpJson.fd, `{
```
Worth a `JSON.stringify({}, null, " ")` to make this a little less awkward?

lukehoban commented on nodejs/eks/cluster.ts:
```diff
@@ -483,14 +485,55 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
+            stdio: "ignore",
+            env: { ...process.env, "KUBECONFIG": tmpKubeconfig.name },
+        });
+    });
```
Minor thing - since we are now doing this in two places - should we extract the common part out into a single function so we can maintain in one place?

metral replied:
Two places? We're only doing this in `createCore()`. What other location are you referring to?

IIUC, we also do `kubectl apply` for VPC if that's what you're referring to, but that is a separate use case from this configmap as we allow user input to be accepted for the VPC.

lukehoban replied:
We do almost the exact same `kubectl apply` in VPC CNI.  Sharing the tmp file creation/writing, processing spawning, etc. in one place seems like it would help us ensure that these are maintained.

lukehoban commented on nodejs/eks/cluster.ts:
```diff
@@ -483,14 +485,55 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
+  }
+},
+"data": ${JSON.stringify(accessData)}
+}`);
```
Is this safe - or might we end up overwriting something already in this `ConfigMap`?  What data does AWS automatically put in here?

metral replied:
> What data does AWS automatically put in here?

AWS automatically puts only the single managed node group role into the `mapRoles` - it does *not* include any additional role mappings, or user mappings for that matter. IMO its which ever nodegroup comes up first that gets added to the config map.

> Is this safe - or might we end up overwriting something already in this ConfigMap?

IME by the time we get to this `kubectl apply` step, the automatic managed group has already occured as it happens behind the scenes. We test this in the new `managed-nodegroups` examples/test with 2 different managed node groups and ensuring we're seeing all expected desired node counts - this is already checked for & actively passing in CI

lukehoban replied:
> IME by the time we get to this kubectl apply step, the automatic managed group has already occured as it happens behind the scenes.

If so - how are we sure we aren't overwriting something that AWS included there?  How do we know that `accessData` includes everything AWS would have included plus everything the user specified?

metral replied:
From the cluster creation POV:

- we have no of knowing this file exists until a `aws.eks.NodeGroup` resource gets created in the cluster, at which point AWS will create it for us and does not inform of us such. We'd have to wait around during cluster creation time, checking for the file to read and parse it if it exists.
- As it stands its only adding the single role for the first managed nodegroup that gets created. Additional node groups don't get their roles added by AWS, so AWS isn't even handling this case correctly when multiple node groups are used - a scenario we currently do handle in EKS correctly

lukehoban replied:
Sorry - I don’t think my question came through clearly - Is it possible that this will ever overwrite (and remove) something that AWS put in place?  Or will this always safely augment what’s already there?  If it is possible for something bad to happen - what is that - and under what conditions?

metral replied:
> Is it possible that this will ever overwrite (and remove) something that AWS put in place

Yes. The `kubectl apply` approach overwrites it with the user and role mappings as provided to the cluster at definition.

Since we have no insight into when AWS creates the configmap, or what it inserts into the configmap (if one is to be created), we have to be signaled somehow to check/read/parse the existing configmap, and then perform some sort of merge with the user provided args. 

Overwriting the configmap with user provided arguments seemed like the easiest path forward to enabling this PR for now, but I'm open to alternatives.

FWIW role and user maps are accepted as input in https://github.com/terraform-aws-modules/terraform-aws-eks/pull/602, but IIUC they don't appear to be currently used or configured on the cluster.

lukehoban commented on nodejs/eks/examples/managed-nodegroups/iam.ts:
```diff
@@ -0,0 +1,27 @@
+        }),
+    });
+
+    let counter = 0;
```
What is `counter` for?

lblackstone replied:
Looks like this creates a unique name for each RolePolicyAttachment that gets generated.

metral replied:
Yes its for unique names. We've used this code before in other examples:

- https://github.com/pulumi/pulumi-eks/blob/master/nodejs/eks/examples/nodegroup/iam.ts
- https://github.com/pulumi/pulumi-eks/blob/master/nodejs/eks/examples/tests/migrate-nodegroups/iam.ts

lukehoban commented on nodejs/eks/nodegroup.ts:
```diff
@@ -288,9 +289,6 @@ export function createNodeGroup(name: string, args: NodeGroupOptions, parent: pu
     }
-    if (core.eksNodeAccess !== undefined) {
-        cfnStackDeps.push(core.eksNodeAccess);
-    }
```
Why is this change made?

metral replied:
`eksNodeAccess` has been replaced with `kubectl apply`, and it was being stored in the `createCore` output `CoreData`. Since we no longer use the CM this dependency is irrelevant and undefined.

lukehoban commented on nodejs/eks/nodegroup.ts:
```diff
@@ -626,3 +624,83 @@ function tagsToAsgTags(tagsInput: InputTags): pulumi.Output<string> {
+     *   - core.privateIds
+     *   - core.publicSublicSubnetIds
+     */
+    subnetIds?: pulumi.Output<pulumi.Output<string>[]>;
```
It's odd that these inputs are of type `pulumi.Output` - are these intended to be `pulumi.Input`?

metral replied:
Done

lukehoban commented on nodejs/eks/nodegroup.ts:
```diff
@@ -626,3 +624,83 @@ function tagsToAsgTags(tagsInput: InputTags): pulumi.Output<string> {
+        subnetIds = core.privateSubnetIds;
+    } else if (core.publicSubnetIds !== undefined) {
+        subnetIds = core.publicSubnetIds;
+    }
```
If none of the above apply - should that be an error instead of falling back to `[]`?

lukehoban replied:
Also - I can't quite reason about whether the above is "right" - is this based on some precedent documented elsewhere either in AWS or in another part of our EKS package?

metral replied:
> If none of the above apply - should that be an error instead of falling back to []?

Good catch. I'll fix

> Also - I can't quite reason about whether the above is "right" - is this based on some precedent documented elsewhere either in AWS or in another part of our EKS package?

This is based on our existing network config setup for self-managed node groups, and using sane defaults OOTB from the cluster object if the user does not specify `subnetIds`.

https://github.com/pulumi/pulumi-eks/blob/79fb423207e0906de815f0d08857f37b85bb14e7/nodejs/eks/nodegroup.ts#L458-L468

Do you think we should not be doing any sort of default logic?

lukehoban replied:
> Do you think we should not be doing any sort of default logic?

I think we should have default logic - I just wasn’t exactly sure how to know whether this specific logic was “right”. If we can cross reference any other source of truth for this - or even just document why we do what we do - that would help future maintainability.

metral replied:
I spell this out in the comments:

https://github.com/pulumi/pulumi-eks/blob/41fc94a183c84d64fe0d0ebb4da368dab2d96bd7/nodejs/eks/nodegroup.ts#L646-L655

>  If we can cross reference any other source of truth for this - or even just document why we do what we do - that would help future maintainability.

This is just repeating the default logic, so I'll go ahead and add a note in the comments about reusing the `computeWorkerSubnet` approach here.

lblackstone commented on nodejs/eks/cluster.ts:
```diff
@@ -483,14 +485,55 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
+}`);
+
+        // Call kubectl to apply the aws-auth manifest.
+        childProcess.execSync(`kubectl apply -f ${tmpJson.name}`, {
```
Just so I'm clear, does this change make a `kubectl apply` call every time `pulumi up` runs? Does that happen both on preview and update?

metral replied:
This does not run on every preview or update. Only in `createCore()` where the configmap was originally being created, and this only runs on initial cluster creation, or when the `nodeAccessData` changes.

I tried moving the configmap creation to both the self-managed and managed node group API calls and those produced unexpected results every time.

metral commented:
I've addressed feedback where possible, but have not yet identified a path forward for how to manage the aws-auth configmap between self-managed and managed node groups.

To summarize what we're working with:

- With self-managed node groups, we will always need to create the configmap as AWS does not create it.
- With managed node groups, IME, only the first nodegroup to connect to the cluster gets its role mapped into the aws-auth configmap, which gets automatically created by AWS. This is not documented anywhere that I've found on TF or AWS.
- Any subsequent managed node groups that are created after the first one is up will **not** be automatically added to the configmap. It appears AWS does not handle this scenario.
  - The proceeding managed node groups register with the cluster, but then become `NotReady` shortly after, and the console shows access denied for the nodes to join the cluster due to their IAM role missing in the configmap.
- The [TF provider](https://www.terraform.io/docs/providers/aws/r/eks_node_group.html) and current [AWS API](https://docs.aws.amazon.com/eks/latest/APIReference/API_Nodegroup.html) does not seem to take into account user and role mappings for the configmap. The only reason we know the aws-auth configmap gets auto-created and populated with the first managed node group's IAM role mapping is because the creation of the `eksNodeAccess` configmap errors with an "existing resource" error via kubectl when we try creating it && using managed node groups simultaneously.
- Given that we don't have a `kubectl apply` or a `get` & `patch` approach (https://github.com/pulumi/pulumi-kubernetes/issues/264) in the event of an existing configmap, and that we encounter [errors](https://github.com/pulumi/pulumi-eks/pull/280#discussion_r352937644) removing the `eksNodeAccess` for existing node group clusters, we must conditionally create or apply the aws-auth configmap based on whether its self or managed node groups. Not to mention, if the user has both self-managed && managed node groups in the same cluster.

The only "clean" way I see of creating the configmap is that we must `get` & `patch` / apply the user args for the configmap w/ any data already in the existing configmap, or create the configmap if it does not exist as usual. But this relies on https://github.com/pulumi/pulumi-kubernetes/issues/264 or some temp work around, and I'm not sure how to proceed.

Thoughts @lukehoban @lblackstone?

lukehoban commented:
I believe that we should be able to keep the original model of creating the `aws-auth` `ConfigMap` ourselves for both self-managed and managed NodeGroups.  We just need to ensure we have a dependency of the `aws.eks.NodeGroup` on the `ConfigMap` to ensure that we always create the `ConfigMap` prior to AWS creating it itself.

metral commented:
Feedback has been addressed and CI is currently running with changes to the nodegroup to now depend on the original `eksNodeAccess` configmap managed by pulumi.

@lukehoban @lblackstone PTAL
